File: backend/examples/file_report_agent.py
import asyncio

from aithru_agent.agent import AgentRuntime, AgentRuntimeResult, PydanticAgentDeps
from aithru_agent.agent.tools import PydanticAIToolBridge
from aithru_agent.application.runtime import create_agent_runtime
from aithru_agent.settings import AgentSettings
from aithru_agent.trace import project_trace_spans


class ToolContext:
    def __init__(self, tool_call_id: str) -> None:
        self.tool_call_id = tool_call_id
        self.run_step = 0
        self.tool_call_approved = False


class FileReportRuntime(AgentRuntime):
    async def run(self, task_msg: str, deps: PydanticAgentDeps) -> AgentRuntimeResult:
        del task_msg
        bridge = PydanticAIToolBridge(deps=deps)
        await bridge.call_tool(
            ToolContext("todo_read_files"),
            "todo.create",
            {"title": "Read files", "status": "running"},
        )
        await bridge.call_tool(
            ToolContext("write_notes"),
            "workspace.write_file",
            {"path": "/inputs/notes.md", "content": "# Notes\nImportant input.\n"},
        )
        await bridge.call_tool(
            ToolContext("read_notes"),
            "workspace.read_file",
            {"path": "/inputs/notes.md"},
        )
        await bridge.call_tool(
            ToolContext("write_report"),
            "workspace.write_file",
            {"path": "/reports/report.md", "content": "# Report\nImportant input.\n"},
        )
        await bridge.call_tool(
            ToolContext("present_report"),
            "presentation.present",
            {"resources": [{"kind": "workspace_file", "path": "/reports/report.md"}]},
        )
        return AgentRuntimeResult(content="Created /reports/report.md")


async def main() -> None:
    runtime = create_agent_runtime(
        settings=AgentSettings(model="test"),
        agent_runtime=FileReportRuntime(),
    )

    run = await runtime.runner.start_run(
        org_id="org_1",
        actor_user_id="user_1",
        task_msg="Analyze workspace files and create a report.",
        scopes=["*"],
    )
    events = await runtime.event_store.list_by_run(run.id)
    spans = project_trace_spans(events)
    files = await runtime.store.list_workspace_files(run.workspace_id)
    report_files = [file for file in files if file.path.startswith("/reports/")]
    report_path = report_files[0].path if report_files else "<none>"
    presentations = [event for event in events if event.type == "presentation.created"]

    print(f"Run id: {run.id}")
    print(f"Run status: {run.status.value}")
    print(f"Report path: {report_path}")
    print(f"Workspace files: {', '.join(file.path for file in files)}")
    print(f"Presentations: {len(presentations)}")
    print(f"Events: {len(events)}")
    print(f"Trace span kinds: {', '.join(sorted({span.kind for span in spans}))}")


if __name__ == "__main__":
    asyncio.run(main())

File: backend/examples/file_report_agent.ts
import { createRuntime } from "@aithru-agent/api";
import type { ToolCallStep } from "@aithru-agent/harness";
import type { AgentRun } from "@aithru-agent/contracts";
import { EVENT_TYPES } from "@aithru-agent/stream";

function now(): string {
  return new Date().toISOString().replace(/\.\d{3}/, "");
}

async function main() {
  const runtime = await createRuntime();

  // Create a run (mimics what API would do)
  const run: AgentRun = {
    id: `run_file_report_${Date.now().toString(36)}`,
    org_id: "org_1",
    actor_user_id: "user_1",
    source: "api",
    thread_id: null,
    workspace_id: `ws_fr_${Date.now().toString(36)}`,
    task_msg: "Analyze workspace files and create a report.",
    scopes: ["*"],
    harness_options: null,
    status: "queued",
    started_at: now(),
    completed_at: null,
    claim: null,
    result: null,
    error: null,
  };
  runtime.store.createRun(run);
  runtime.eventWriter.write(
    run.id,
    run.thread_id ?? null,
    EVENT_TYPES.RUN_CREATED,
    { run_id: run.id, status: run.status },
  );

  // Define the scripted steps (same as Python FileReportRuntime)
  const script: ToolCallStep[] = [
    {
      name: "todo.create",
      input: { title: "Read files", status: "running" },
    },
    {
      name: "workspace.write_file",
      input: { path: "/inputs/notes.md", content: "# Notes\nImportant input.\n" },
    },
    {
      name: "workspace.read_file",
      input: { path: "/inputs/notes.md" },
    },
    {
      name: "workspace.write_file",
      input: { path: "/reports/report.md", content: "# Report\nImportant input.\n" },
    },
    {
      name: "presentation.present",
      input: {
        resources: [{ kind: "workspace_file", path: "/reports/report.md" }],
      },
    },
  ];

  // Execute
  const completedRun = await runtime.worker.startRun(run, {
    steps: script,
    finalContent: "Created /reports/report.md",
  });

  // Gather results
  const events = runtime.store.listEvents(run.id);
  const files = runtime.store.listWorkspaceFiles(run.workspace_id);
  const reportFiles = files.filter((f) => f.path.startsWith("/reports/"));
  const reportPath = reportFiles.length > 0 ? reportFiles[0].path : "<none>";
  const presentations = events.filter(
    (e) => e.type === "tool.completed" && (e.payload as any)?.name === "presentation.present",
  );

  // Print results (same shape as Python output)
  console.log(`Run id: ${completedRun.id}`);
  console.log(`Run status: ${completedRun.status}`);
  console.log(`Report path: ${reportPath}`);
  console.log(
    `Workspace files: ${files.map((f) => f.path).join(", ") || "(none)"}`,
  );
  console.log(`Presentations: ${presentations.length}`);
  console.log(`Events: ${events.length}`);

  // Print event types
  const eventTypes = [...new Set(events.map((e) => e.type))].sort();
  console.log(`Event types: ${eventTypes.join(", ")}`);

  // Verify acceptance criteria
  const requiredEvents = [
    "run.created",
    "run.started",
    "message.created",
    "message.delta",
    "message.completed",
    "run.completed",
  ];
  const missing = requiredEvents.filter((t) => !eventTypes.includes(t));
  if (missing.length > 0) {
    console.log(`\nWARNING: Missing required events: ${missing.join(", ")}`);
  } else {
    console.log(`\nAll required events present.`);
  }
}

main().catch((err) => {
  console.error("File report agent failed:", err);
  process.exit(1);
});
